### ISSUE TITLE
3.6.5在模拟器上测试报错

### ISSUE THREAD

**Issue by sisier** [hannesa2/paho.mqtt.android#699]
Process: com.ceshi.demo, PID: 8808
                                                    java.lang.RuntimeException: Error receiving broadcast Intent { act=.callbackToActivity.v0 flg=0x10 pkg=com.ceshi.demo(has extras) } in info.mqtt.android.service.MqttAndroidClient@2bbad63
                                                    	at android.app.LoadedApk$ReceiverDispatcher$Args.lambda$getRunnable$0(LoadedApk.java:1822)
                                                    	at android.app.LoadedApk$ReceiverDispatcher$Args.$r8$lambda$mcNAAl1SQ4MyJPyDg8TJ2x2h0Rk(Unknown Source:0)
                                                    	at android.app.LoadedApk$ReceiverDispatcher$Args$$ExternalSyntheticLambda0.run(D8$$SyntheticClass:0)
                                                    	at android.os.Handler.handleCallback(Handler.java:959)
                                                    	at android.os.Handler.dispatchMessage(Handler.java:100)
                                                    	at android.os.Looper.loopOnce(Looper.java:232)
                                                    	at android.os.Looper.loop(Looper.java:317)
                                                    	at android.app.ActivityThread.main(ActivityThread.java:8705)
                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:580)
                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:886)
                                                    Caused by: java.lang.ClassCastException: org.eclipse.paho.client.mqttv3.MqttException cannot be cast to android.os.Parcelable
                                                    	at info.mqtt.android.service.MqttAndroidClient.connectionLostAction(MqttAndroidClient.kt:1304)
                                                    	at info.mqtt.android.service.MqttAndroidClient.onReceive(MqttAndroidClient.kt:905)
                                                    	at android.app.LoadedApk$ReceiverDispatcher$Args.lambda$getRunnable$0(LoadedApk.java:1814)
                                                    	at android.app.LoadedApk$ReceiverDispatcher$Args.$r8$lambda$mcNAAl1SQ4MyJPyDg8TJ2x2h0Rk(Unknown Source:0) 
                                                    	at android.app.LoadedApk$ReceiverDispatcher$Args$$ExternalSyntheticLambda0.run(D8$$SyntheticClass:0) 
                                                    	at android.os.Handler.handleCallback(Handler.java:959) 
                                                    	at android.os.Handler.dispatchMessage(Handler.java:100) 
                                                    	at android.os.Looper.loopOnce(Looper.java:232) 
                                                    	at android.os.Looper.loop(Looper.java:317) 
                                                    	at android.app.ActivityThread.main(ActivityThread.java:8705) 
                                                    	at java.lang.reflect.Method.invoke(Native Method) 
                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:580) 
                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:886) 

.6.5就报错，页面长时间静止下，过一会就崩了。。在3.6.4没遇到问题

**Comment by hannesa2** [comment]
Version 3.6.4 and MqttAndroidClient.kt does not come with a line 1304 https://github.com/hannesa2/paho.mqtt.android/blob/3.6.4/serviceLibrary/src/main/java/info/mqtt/android/service/MqttAndroidClient.kt

Without the chance to reproduce it, I'm helpless here

**Comment by hmueller01** [comment]
Same here
```
 FATAL EXCEPTION: main (Ask Gemini)
                                                                                                    Process: de.euhm.solar, PID: 31426
                                                                                                    java.lang.RuntimeException: Error receiving broadcast Intent { act=.callbackToActivity.v0 flg=0x10 pkg=de.euhm.solar (has extras) } in info.mqtt.android.service.MqttAndroidClient@1d342b5
                                                                                                    	at android.app.LoadedApk$ReceiverDispatcher$Args.lambda$getRunnable$0(LoadedApk.java:1822)
                                                                                                    	at android.app.LoadedApk$ReceiverDispatcher$Args.$r8$lambda$mcNAAl1SQ4MyJPyDg8TJ2x2h0Rk(Unknown Source:0)
                                                                                                    	at android.app.LoadedApk$ReceiverDispatcher$Args$$ExternalSyntheticLambda0.run(D8$$SyntheticClass:0)
                                                                                                    	at android.os.Handler.handleCallback(Handler.java:959)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:100)
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:232)
                                                                                                    	at android.os.Looper.loop(Looper.java:317)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:8705)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:580)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:886)
                                                                                                    Caused by: java.lang.ClassCastException: org.eclipse.paho.client.mqttv3.MqttException cannot be cast to android.os.Parcelable
                                                                                                    	at info.mqtt.android.service.MqttAndroidClient.connectionLostAction(MqttAndroidClient.kt:1304)
                                                                                                    	at info.mqtt.android.service.MqttAndroidClient.onReceive(MqttAndroidClient.kt:905)
                                                                                                    	at android.app.LoadedApk$ReceiverDispatcher$Args.lambda$getRunnable$0(LoadedApk.java:1814)
                                                                                                    	at android.app.LoadedApk$ReceiverDispatcher$Args.$r8$lambda$mcNAAl1SQ4MyJPyDg8TJ2x2h0Rk(Unknown Source:0) 
                                                                                                    	at android.app.LoadedApk$ReceiverDispatcher$Args$$ExternalSyntheticLambda0.run(D8$$SyntheticClass:0) 
                                                                                                    	at android.os.Handler.handleCallback(Handler.java:959) 
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:100) 
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:232) 
                                                                                                    	at android.os.Looper.loop(Looper.java:317) 
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:8705) 
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method) 
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:580) 
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:886) 
```

It happens after I set `targetSdk 34`.

Upgrade Assisant says:

> For apps targeting Android 14, Android restricts apps from sending implicit intents to internal app components in the following ways:
> Implicit intents are only delivered to exported components. Apps must either use an explicit intent to unexported components, or mark the component as exported.
> If an app creates a mutable pending intent with an intent that doesn't specify a component or package, the system now throws an exception.
> These changes prevent malicious apps from intercepting implicit intents that are intended for use by an app's internal components.

**Comment by hannesa2** [comment]
You could give https://github.com/hannesa2/paho.mqtt.android/releases/tag/4.3.beta5 a try

**Comment by hmueller01** [comment]
The RuntimeException is gone. But still not working. It says the schema is missing. With 3.x I did not use a schema. Just the server name and the port. What schema should I use for a TLS connection on port 8883? ssl:// tls:// mqtts://?

And there is an issue with
```
		mMqttAndroidClient = new MqttAndroidClient(mContext, brokerUrl, clientId, Ack.AUTO_ACK);
```
it says _Expected 7 arguments but found 4_.
Using 
```
		mMqttAndroidClient = new MqttAndroidClient(mContext, clientId, brokerUrl, Ack.AUTO_ACK, null, false, 1000);
```
it works, but I do not understand why Java complains, because these should be optional parameter ...

**Comment by hannesa2** [comment]
@hmueller01 
Without a sample, which demonstrate it, I've no clue

**Comment by hannesa2** [comment]
@hmueller01
> it says Expected 7 arguments but found 4.

> I do not understand why Java complains, because these should be optional parameter ...

The default parameter doesn't work with Java. My recommendation: use Kotlin !

**Comment by hmueller01** [comment]
public static void validateURI(String brokerUri) throws IllegalArgumentException {
		try {
			URI uri = new URI(brokerUri);
			String scheme = uri.getScheme();
			if (scheme == null || scheme.isEmpty()) {
				throw new IllegalArgumentException("missing scheme in broker URI: " + brokerUri);
			}


> @hmueller01 Without a sample, which demonstrate it, I can't do anything here

Sorry.

It's in org.eclipse.paho.client.mqttv3-1.2.5.jar
```
	public static void validateURI(String brokerUri) throws IllegalArgumentException {
		try {
			URI uri = new URI(brokerUri);
			String scheme = uri.getScheme();
			if (scheme == null || scheme.isEmpty()) {
				throw new IllegalArgumentException("missing scheme in broker URI: " + brokerUri);
			}
```

**Comment by hmueller01** [comment]
This is weird. All of a sudden the exception is gone.
I just added a output of the serverURI. It is ssl://broker:8883
But still I do not get any message. I think this is because things changed in 4.3 compared to 3.6. Is there a documentation what I have to do in detail?

**Comment by hannesa2** [comment]
Without running sample, it smells like the issue is on your side. 
You don't need much to do on upgrade https://github.com/hannesa2/paho.mqtt.android?tab=readme-ov-file#version-4x-doesnt-need-a-foreground-service

**Comment by hmueller01** [comment]
> @hmueller01
> 
> > it says Expected 7 arguments but found 4.
> 
> > I do not understand why Java complains, because these should be optional parameter ...
> 
> The default parameter doesn't work with Java. My recommendation: use Kotlin !

I am so stupid. You are right. Java has no optional parameter. This has to be implemented by overloading the method. I was used the Kotlin way from Python or C++. So yeah, Kotlin is so much better than Java. Porting my app might be a good idea on the long run.

**Comment by hmueller01** [comment]
I was just trying to switch back to 3.6 but this fails at backporting the RoomDatabase ...
```
FATAL EXCEPTION: DefaultDispatcher-worker-2 (Ask Gemini)
                                                                                                    Process: de.euhm.solar, PID: 13949
                                                                                                    java.lang.IllegalStateException: A migration from 2 to 1 was required but not found. Please provide the necessary Migration path via RoomDatabase.Builder.addMigration(Migration ...) or allow for destructive migrations via one of the RoomDatabase.Builder.fallbackToDestructiveMigration* methods.
                                                                                                    	at androidx.room.RoomOpenHelper.onUpgrade(RoomOpenHelper.kt:110)
                                                                                                    	at androidx.room.RoomOpenHelper.onDowngrade(RoomOpenHelper.kt:123)
                                                                                                    	at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.onDowngrade(FrameworkSQLiteOpenHelper.kt:277)
                                                                                                    	at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:413)
                                                                                                    	at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:316)
                                                                                                    	at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.getWritableOrReadableDatabase(FrameworkSQLiteOpenHelper.kt:232)
                                                                                                    	at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.innerGetDatabase(FrameworkSQLiteOpenHelper.kt:190)
                                                                                                    	at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.getSupportDatabase(FrameworkSQLiteOpenHelper.kt:151)
                                                                                                    	at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper.getWritableDatabase(FrameworkSQLiteOpenHelper.kt:104)
                                                                                                    	at androidx.room.RoomDatabase.inTransaction(RoomDatabase.kt:638)
                                                                                                    	at androidx.room.RoomDatabase.assertNotSuspendingTransaction(RoomDatabase.kt:457)
                                                                                                    	at info.mqtt.android.service.room.MqMessageDao_Impl.allArrived(MqMessageDao_Impl.java:318)
                                                                                                    	at info.mqtt.android.service.MqttConnection$deliverBacklog$1.invokeSuspend(MqttConnection.kt:222)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
                                                                                                    	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:115)
                                                                                                    	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:100)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684)
                                                                                                    	Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@846ffb9, Dispatchers.IO]
```

Which can be solved only by cleanup the storage memory. And this is bad. Could you please implement the necessary migration.

**Comment by hannesa2** [comment]
How should this work? 
Upgrade migration is a common task.
Downgrade is an individual task. From where to what?

**Comment by hmueller01** [comment]
Sorry, but the Exception is there again. 

```
	public void connect(String brokerUrl,
						String user, String pass,
						int securityMode, String caFile,
						String clientId) {
		Log.d(LOG_TAG, "brokerUrl: " + brokerUrl);
		//mMqttAndroidClient = new MqttAndroidClient(mContext, brokerUrl, clientId, Ack.AUTO_ACK);
		mMqttAndroidClient = new MqttAndroidClient(mContext, clientId, brokerUrl, Ack.AUTO_ACK, null, false, 1000);
		IMqttToken token = mMqttAndroidClient.connect(getMqttConnectionOption(user, pass, securityMode, caFile));
		token.setActionCallback(new IMqttActionListener() {
			@Override
			public void onSuccess(IMqttToken asyncActionToken) {
				mMqttAndroidClient.setBufferOpts(getDisconnectedBufferOptions());
				Log.d(LOG_TAG, "onSuccess(): Successfully connected");
			}

			@Override
			public void onFailure(IMqttToken asyncActionToken, final Throwable exception) {
				Log.d(LOG_TAG, "onFailure(): " + exception.toString());
				Handler handler = new Handler(Looper.getMainLooper());
				handler.post(new Runnable() {
					@Override
					public void run() {
						Toast.makeText(mContext,
								mContext.getResources().getString(R.string.mqtt_connect_failed) +
										exception.toString(), Toast.LENGTH_LONG).show();
					}
				});
			}
		});
	}
```

It jumps to onFailure with
`
2024-10-26 11:41:22.682  4742-4827  MqttClient              de.euhm.solar                        D  onFailure(): java.lang.IllegalArgumentException: missing scheme in broker URI: 1b1d1a2428457128`

Searching for the "missing scheme in broker URI" I find that it comes from org.eclipse.paho.client.mqttv3-1.2.5.jar in NetworkModuleService.java
```
public static void validateURI(String brokerUri) throws IllegalArgumentException {
		try {
			URI uri = new URI(brokerUri);
			String scheme = uri.getScheme();
			if (scheme == null || scheme.isEmpty()) {
				throw new IllegalArgumentException("missing scheme in broker URI: " + brokerUri);
			}
			scheme = scheme.toLowerCase();
			synchronized (FACTORY_SERVICE_LOADER) {
				for (NetworkModuleFactory factory : FACTORY_SERVICE_LOADER) {
					if (factory.getSupportedUriSchemes().contains(scheme)) {
						factory.validateURI(uri);
						return;
					}
				}
			}
			throw new IllegalArgumentException("no NetworkModule installed for scheme \"" + scheme
					+ "\" of URI \"" + brokerUri + "\"");
		} catch (URISyntaxException e) {
			throw new IllegalArgumentException("Can't parse string to URI \"" + brokerUri + "\"", e);
		}
	}
```

As said, my URI is 
`brokerUrl: ssl://broker:8883` 
and it seem like this is not passed to paho ...

**Comment by hannesa2** [comment]
yours: **1b1d1a2428457128**

> missing scheme in broker URI: 1b1d1a2428457128

compared to my `extendedSample` 

<img width="809" alt="image" src="https://github.com/user-attachments/assets/01a40de4-3caf-4688-a69a-e688585df453">

You spam this Chinese (!) ticket, where I can't read Chinese with your beginner and local issues. 
I asked for a running sample, to demonstrate issue. -Just ignored-
What's going on here ? 

Please open an own ticket for this.

**Comment by hmueller01** [comment]
Sorry for "_spamming_" here. It just started that v3.6.5 won't run with targetSdk > 33. So I found this issue as I saw the same 

> java.lang.RuntimeException: Error receiving broadcast Intent

You suggested using v4.3.beta5. I tried to test it and just commented here my issues I found with that. Sorry that I am not a profi programmer that finds all bugs within a second. But with your indirect hint I found it! Thanks a lot!

I do not know how this happend, but while copying 
`mMqttAndroidClient = new MqttAndroidClient(mContext, clientId, brokerUrl, Ack.AUTO_ACK, null, false, 1000);`
I mixed the order of `clientId` and `brokerUrl`.

I should have found this without asking here, as
`throw new IllegalArgumentException("missing scheme in broker URI: " + brokerUri);`
should have given me the exact hint. So sorry again!

At the end of the day v4.3.beta5 works with targetSdk 34 and 35. Closing here.
